### ISSUE TITLE
maxsat interface á la Prover interface

### ISSUE THREAD

**Issue by frehn** [gapt/gapt#308]
Having an abstract interface for "provers", i.e. the Prover trait, has been very useful so far since provers can be easily implemented independently from each other, and plugged into algorithms using them (e.g. CutIntro).

For "MaxSAT solvers", such an interface does not exist yet. I think one should come up with a trait MaxSATSolver and refactor the current MaxSAT.scala into 4 classes (QMaxSAT, ToySAT, ToySolver, MiniMaxSAT) implementing that trait.

An initial suggestion for the trait would be to simply generalize the solve method from the current implementation.

```scala
trait MaxSATSolver {
  /**
   * @param hard Hard constraints.
   * @param soft Soft constraints in CNF along with their weights.
   * @return None if hard is unsatisfiable, otherwise Some(model), where model is a model
   * of hard maximizing the sum of the weights of soft.
   */
  def solve( hard: List[FClause], soft: List[(FClause, Int)] ): Option[Interpretation]
}
```

Input from people familiar with the formal definition of MaxSAT, and the implementation of MaxSAT in GAPT, is of course very much appreciated :-)

**Comment by DarwinsBuddy** [comment]
This seems to be a right design decision.
Currently the MaxSAT interface is more like a factory wrapper.
In solve we would have to handle of course in/output of various already
supported MaxSATSolvers.
Please be aware of the fact, that although all of them stating to use the
DIMACS .wcnf format, they do not really implement the standard definition
of DIMACS Format
<http://www.logoptimize.it/index.php/menu-input-format/menu-dimacs-input>.
Therefore I had to write various in/output parsing functions to obtain
interpretations and pass formulas.


2015-03-19 15:09 GMT+01:00 Daniel Weller <notifications@github.com>:

> Having an abstract interface for "provers", i.e. the Prover trait, has
> been very useful so far since provers can be easily implemented
> independently from each other, and plugged into algorithms using them (e.g.
> CutIntro).
>
> For "MaxSAT solvers", such an interface does not exist yet. I think one
> should come up with a trait MaxSATSolver and refactor the current
> MaxSAT.scala into 4 classes (QMaxSAT, ToySAT, ToySolver, MiniMaxSAT)
> implementing that trait.
>
> An initial suggestion for the trait would be to simply generalize the
> solve method from the current implementation.
>
> trait MaxSATSolver {  /**   * @param hard Hard constraints.   * @param soft Soft constraints in CNF along with their weights.   * @return None if hard is unsatisfiable, otherwise Some(model), where model is a model   * of hard maximizing the sum of the weights of soft.   */
>   def solve( hard: List[FClause], soft: List[(FClause, Int)] ): Option[Interpretation]
> }
>
> Input from people familiar with the formal definition of MaxSAT, and the
> implementation of MaxSAT in GAPT, is of course very much appreciated :-)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/gapt/gapt/issues/308>.
>



-- 
Tel.: +43 676 643 00 55
EMail: christoph.spoerk@gmail.com

**Comment by frehn** [comment]
@DarwinsBuddy: Thanks for the comment!

Fixed in 2c1bb946489b2517852c3614b3e8de81c08aa2a9.
